### ISSUE TITLE
Use of 'api' in HardwareRepository

### ISSUE THREAD

**Issue by rhfogh** [mxcube/HardwareRepository#366]
Ivars pointed out that one should not use api in the HardwareRepository, since it is out-of-repository. I can see the point, but I had stared already, since 'api' solved a problem I had met a lot in my HOs. If you need access to a HardwareObject from another you get into a lot of mess since you cannot really predict which one is loaded at which point, so you cannot set it from the init of your own HO, or, easily, from its properties. 

Basically you need that central location where you can get hold of all the HardwareObjects, and you need it accessible also from the HardwareObjects. 

The current API only really contains links to HardwareObjects anyway. Could we move it into the HarwdwareRepository, and use the (top-level) api for other things? Or what shoudl we do?

**Comment by marcus-oscarsson** [comment]
I'm happy to see that the API discussion continues :), 

First of all I think we need to establish what we mean by API in this case, a central location of where one access an arbitrary HO is not necessarily an API. We need to define the exact interface and behavior of that object before it becomes an API. A common beamline or application object that provides a convenient access point to the various parts of the application is certainly useful and we have so far been discussing this as the "Beamline" object, partly replacing the existing BeamlineSetup. 

What we have been discussing is an API between the UI and backend. Its interface would naturally therefore reflect the needs of a client application of HardwareRepoistory rather than the internals of HardwareReopistory. You are maybe working around a different problem by using the "API" from within HardwareRepository ? We can of course discuss the usage of this API within the HardwareRepository as well but I'm not sure it would really solve your problem ?

**Comment by rhfogh** [comment]
What I need for now is a beamline object, somewhere central I can get hold of all the hardware objects without worrying about property names or loading order. And the 'api' as currently coded is exactly that. Which is why I started using it. Indeed I am grateful to Ivars for doing the work of making it, but I think it is misnamed and misplaced for what it actually does. The beamline object naturally belongs in the HardwareObjects repository, so you can also access it from there. I would argue that the actual 'api' object should also live in the HardwareRepository. Actually, 'api' stands for 'Application Program *Interface*', and if it is the interface to the HardwareObjecst layer it belongs in the HardwareObjects layer. Of course that depends on what functionality you want it to contain. But one way or the other it would be silly to have all these well-organised functions to access the HO layer and then make it impossible to use from withihin the HO layer.

Should we discuss whether to rename, move, and/or split the current 'api'?

**Comment by marcus-oscarsson** [comment]
I agree on that this API belongs to the HardwareRepoistory, I dont see why we would need to rename, split or move the API from HardwareReopistory. Its intention is to define a interface between the UI/Client and the backend. The client could possibly be the HardwareRepository itself but I'm wondering if you are not solving a different problem by using it that way ?. As I understand it you would like to ensure that the HO's you have configured are actually correctly loaded ? Would not this be a function of the HardwareRepository itself ?

**Comment by rhfogh** [comment]
The problem is not how to get my own HO to load, but to set up links from them to other HO without anything breaking. I have had two problems that the 'api' solves:

First, I need to get hold of various other HOs from inside mine. I do not want to put e.g. 'collect' as a property of the GPhL hardwareObjects, because the properties should be specified in only one place. I do not always have a handle to another object that  can use to find them, and using getHardwareRepository()... is clunky and puts the property name strings deep in the bowels of the program. The 'api' is a better solution.

Second, if I need a handle on e.g. 'collect' or 'queue' in several functions, the obvious way is to set it up in the __init__ of my own HO. Here I get a lot of trouble, because that only works if the other HO are fully loaded by the time my HO starts loading, which requires either careful attention to the load sequence (that is not precisely defined anywhere) or setting the relevant 'collect_hwobj' attribute from some kind of set_up function that must be called after the object has finished initialising, but before I actually start executing it. I have spent a certain amount of time fiddling with these things.

Surely the solution is to have a single object that can be imported and called at run time and that is known to have a link to all the HO? Then you can simply import that and access it at need at runtime, without bothering about setting links to other HO as attributes. I would call it a beamline object rather than an api, myself. But what I need corresponds exactly to the contents of the current api.__init__

BTW, are you aware that the api currently does *not* live in the HarwareRepository, but in the mxcube repository?

**Comment by marcus-oscarsson** [comment]
The way to do what you are describing is currently via properties or via getHardwareObject(). We have discussed an object that centralizes access to other HardwareObjects and we have refereed to this idea as Beamline or Application object. This object can of course be used within the HardwareRepository and part of it might be part of the interface between the UI and the backend.

I see the API that you are currently using more as an implementation idea or proof of concept of the UI API that we have been discussing. Its not something that we have agreed on yet. I think we are getting a bit to eager here, the discussion is good and it really highlights the two different needs.

1. Need for the UI-API as we said on the last meeting, this needs further discussion
2. Need for common Beamline object

**Comment by marcus-oscarsson** [comment]
The API you are using is also only available in the Qt version so what you are doing would not be compatible with the web version ...

**Comment by rhfogh** [comment]
@marcus-oscarsson 
I do not quite understand what we are disagreeing about:
- We all agree that there is a need for a beamline object (it  is in the roadmap - Milestone 3). 
- We agree that it should contain links to the HardwareObjects (like 'api' does now)
- We agree that it should be located in the HardwareRepository
- We agree, apparently, that the experimental 'api' that Ivars wrote is a good thing, and that it is fine for him to be coding to it for now - also in advance of any thorough discussion or agreement on the matter. 
- We agree that using getHardwareObject is not a good approach. In addition to what I said already, I would point out that getHardwareObject is keyed to the names of .xml config files, but that the names or locations of the config files are not standardised across beamlines. You cannot reliably get hold of the local HardwareObject from your code unless you know what the local file name is.

Given all that agreement, I admit I am reluctant to spend time changing my code back to its previous, more complex state and debugging the result, in order to comply with the 'current way to do what I am describing'. 

I would therefore propose a compromise solution:

- I write a draft Beamline object and put it in the HardwareObjects repository
- The contents will be an exact duplicate of Ivars api, since that is what we seem to need
- The file will be called GphlBeamline, to make it clear that this is GPhL-specific code, and so that other people do not get misled into using it.
- I use the GphlBeamline object for access to HardwareObjects in GPhL code for now, but promise to change over to whatever implementation we agree on as soon as that implementation is in place.

How does that sound?

**Comment by marcus-oscarsson** [comment]
Oh, I don't think we were actually disagreeing about anything I simply tried to answer your question. The answer was simply that I don't think its a good idea to use the API from the mxcube repository in your HardwareRepository code especially not if you at some point want it to be compatible with MXCuBE3. I think we agree on the rest.

I think that your compromise solution is fine if we first define a common Beamline object and discuss it through a PR in the lines with the refactoring work that we are currently doing. You can then derive a GphlBeamline from that if you have some specific needs that are not full filled by the common one.

As a side note I have myself also started to define a beamline object for BSXCuBE that is used from another application wide object called Application. I would be happy to discuss and define a Beamline object that we can reuse in that application as well.

**Comment by rhfogh** [comment]
@marcus-oscarsson 
"_first define a common Beamline object and discuss it through a PR in the lines with the refactoring work that we are currently doing_"
Excellent! That is exactly what I wanted. Preferably we should start immediately, since my last check-in goes against the current guidelines, and I would quite like to fix that as soon as possible. 

I would be happy to come up with a first draft in a day or two.Could you send me a link or draft or something to the beamline object you are making for the BSXCuBE? Then I can make a draft that fits with that.

BTW it is highly unlikely that I would ever need a GPhL-specific object. The only reason I proposed it is that it looked like the only way to get something started now. I had the impression from your previous comments that you would be unwilling to accept  anything more general than that pending a future reorganisation and agreement.

**Comment by marcus-oscarsson** [comment]
@rhfogh 

Sure the sooner the better :), the one I'm currently working on can be found here

    https://github.com/bsxcube/bsx3/tree/master/external/SAXHO

Its for the moment quite limited. The idea is simply to define an object that gives access to a certain set of HardwareObjects and routines through a well defined interface. I've started to make it possible to specify if a HardwareObject is optional or not. I was then thinking that it would be better to extend HardwareRepository to be able to handle that "natively". The application simply exits with en error message if a non optional HardwareObject cant be initialized.

The beamline object itself is available trough a globally available Application object, which one for the moment access trough a get_app() method. One simply does:

    get_app().beamline.energy

to get a hold of the energy HardwareObject for instance.

**Comment by rhfogh** [comment]
Thanks. I shall try to check in something by the end of the week.
I agree that it is better to have an object than (as in Ivars' api) a module. Not sure whether this might after all be a good place to specify which objects are mandatory - that means that beamline owners and programmers both have a reference. Anyway, I shall look.

Where does your get_app() live? Is it part of the UI or the HardwareRepository? One would clearly like the Beamline object to be a singleton, so the question is where that should be handled.

**Comment by marcus-oscarsson** [comment]
It could be part of the HardwareRepoistory and I have been thinking about bringing the idea up with you guys. It currently lives in the backend of the application, not the UI part.

    [https://github.com/bsxcube/bsx3/blob/master/bsx3/backend/bsxapp/__init__.py](https://github.com/bsxcube/bsx3/blob/master/bsx3/backend/bsxapp/__init__.py)

**Comment by marcus-oscarsson** [comment]
The Beamline object would be a singleton automatically if it's made a HardwareObject itself which is not really necessary but its how I did it for the moment. I'm still working on this so my idea is not completely finished yet and I were going to make a PR at some stage. But you can go ahead and do something if you are motivated :)

**Comment by rhfogh** [comment]
I think it would be important that you can get hold of it by import, i.e. without needing an object that you can call getHardwareObject on. But I shall see what I can come up with.
